### PR TITLE
Template.contentBlock and @index

### PR DESCRIPTION
**Pull request by faceyspacey**
i used `\s*` instead of `\s+` because spaces arent required. ...also note: this won't actually be encountered until we build the lookup that handles templates used as a helper, i.e. the place where `contentBlock` comes from.

### CODE REVIEW

File: react-regex.js [hunk 1]
```diff
@@ -174,5 +174,15 @@ ReactRegex = [
...
+  },
+  // Template.contentBlock
+  {
+    regex: /({{>\s*Template.contentBlock\s*}})/g,
+    replace: "{this.props.children}"
+  },
+  // @index -- match @index by itself {{@index}} or as a param to helper {{helper @index anotherParam}}
+  {
+    regex: /({{.*@index.*}})/g,
+    replace: "{index}"
```
Comment (timbrandin) [inline react-regex.js:186]: I'm haven't tested this, but it feels like this would replace the whole `{{helper @index anotherParam}}` with {index}, right?
http://regexr.com/3c6ap

Comment (faceyspacey) [inline react-regex.js:186]: yea, the way you're injecting variables such as `context` (and `index` from `each in`) into scope will make this work as long as something else later injected doesnt override the scope/variable. 

ps. we need to get some support on this project. Look, you were discussed in depth in Josh Owen's latest podcast: https://www.youtube.com/watch?v=BCNEmB3mI78 ...It was near the beginning (within the first 15 I recall). Sideburns was literally one of the 3-4 long-form pre-scheduled segments on the show! This is a hot project. Lots of people are gonna benefit from it. We need to find a way to get some of these guys to help. I know there are a few guys that will pick it up and run with it hard within their own packages. So they should be helping us so we can all get to that point quicker.

Comment (timbrandin) [inline react-regex.js:186]: Wow I didn't notice they talked about it. This is supercool! Yeah, we need to spread this project even more, but we need to become fully Blaze 1 compatible first?

Comment (faceyspacey) [inline react-regex.js:186]: hmmm, not to get help. Im busy as hell just like you. But what would do, and will do as soon as i really dive in, is refactor it, and document the flow visually and then send out the readme page that contains the diagram and contribution notes to likely contributors and essentially request their help. We just gotta come correct. Right now if we send it out to 10-20 devs I know that would be willing to work on it, they don't have much to go by. It took me many hours just trying to figure it out. Also, having the tests setup as part of that would be a great way to make your checklist easy to work on, since the tests would be a guide for exactly what they need to work on and serve as great satisfaction if they are able to get even just one yet-to-be-developed failing test to pass. 

It's all about marketing. In short we could spend more time doing it all ourselves or a lot less time on marketing. We need to make it as easy as possible to contribute. So our check list is:
- refactor strategy
- diagram of new architecture
- integration tests (stuff like in @TwinTailsX 's directory)
- spacebars unit tests
- demo repo showing what does work (again, basically stuff in @TwinTailsX 's directory, but more)
- OPTIONAL: a few more things working: template inclusion, regular #each, helper data context, events data context

....the basic idea being that what takes the most time getting to work is stuff in the OPTIONAL category and beyond, the actual software. But refactoring, and diagraming out how it works (and believe me we need this! its abnormal programming what we're doing, so we need to make it as clear as possible), and at least some basic tests are quick relatively painless endeavor. If it's taking a while, we dont need to have that many tests and we dont need to do much more on the actual software. As far as the software, we just need to make sure that what we do have works, so developers can load up the repo and shit aint broken. What has been checked passes its tests in other words. 

Once we get this "marketing kit" together we post on the Meteor Forums and crater, etc. ..I mean you saw how much activity was in that Blaze thread! This will likely garner a lot of interest, and a few guys will step up to the plate. Anybody who's clamoring for this, we put them to work--instead of letting them wait around for somebody else (us) to do it. 

To me the most important part is documenting out the general flow of how it currently works and then come up with a better way to do it. Jumping back and forth between 2 repositories doesn't seem like a good thing at this stage. I vote to combine them. I also think u should rename the project. Sideburns has no meaning to those not in the "know", i.e. those that haven't heard about the package yet. Its name needs to convey its meeting. eg: Blaze React, Blaze React Renderer. github.com/timbrandin/blaze-react should probably just be it. ..So we got it all combined into one new project called Blaze React--we next need to figure out if there's a better flow than essentially doing endless passes over spacebars code, injecting variables into scope. I'm not saying we dont do that--i dont think we have any choice, but I could be wrong. I'm saying that the way it currently works is non-obvious: first u do events in a totally separate compiler, then u have those events available to the html compiler, which u handle first; then u convert helpers, then in a non standard way u do this `toString()` stuff (i.e: `source = "" + new ReactTemplate(original)`) which i think should be changed to a more standard method call (it made me unsure what this was at first); then u finally get back to parsing everything, at which point u handle templates. I.e. u move from inner to outer. If it's possible we should do it so u first parse the templates, and the parser naturally recurses into the code it finds. 

Going from events (which seems almost unrelated to html) then to helpers, and then back up to templates seems unnatural to me. The next challenging part is not only is the html being converted, but then u have functions in the code that do even more stuff. So then u gotta follow those functions, and now ur lost! We should simplify this all as soon as possible before the technical debt amounts. The functions in the converted code are likely unavoidable, but the plugin parsers up until that point need to be completely clear and clean. One thing we may wanna do is convert the plugin into one that handles js and html. I forget if they can handle more than one extension, but if we can, we should, and then in our own code we diverge off into the 2 things (events + helpers/templates). That way you can look at the code and sequentially scan down it while seeing that events are handled first, then templates, then helpers, or whatever order we end up coming up with. THEN, all the code related to standard transpilation should be moved to another file and called form this main plugin file. Perhaps we even have 3 files: 1) a near-empty main plugin file, which calls code from the 2) standard Babel transpilation file and 3) from our Blaze-React specific class file. That will do wonders to put all our core unique transpilation logic in one logical sequentially understandable place, unobstructed by unrelated code. 

NEXT: the code in the exports.js file gets killed and moved to what's in template.jsx in the current main sideburns repo.  Of course now we're all in one repo, but nevertheless RT is in template.jsx part of the react component class. After all, the code is being called in a react render method anyway. But perhaps more importantly--and this could be a game changer--those RT methods like `lookup()` now get access to `this.props/state/etc`. So now your parser doesn't have to inject as much stuff. `this.props` should be the context always, unless in `#each/with/etc` sections. NEXT: helpers should be attached to that class as well. You're currently doing a lot of hard to understand work here: https://github.com/timbrandin/sideburns/blob/devel/template.jsx#L23 

It seems to me that if we just moved the methods from helpers to the react component, they'd just be available, and we move the work of setting the context to a single place. Currently setting the context is split between there and your RT exports and maybe other places. Let's just move the methods from the blaze helpers map to the react component. In the transpiled code we have: `this.lookup('helperName')` must like Blaze does, and then we leave the context discovery all to the `lookup` methods. Now we dont need to worry about where the context is being handled among 2+ places. 

Events too should be handled the same way: an event called: 'click .someClass' becomes a method on the component of the same name. And then when we encounter that selector in the html, we simply assign: `<div className="someClass" onClick={this['click .someClass']}>` . For now we won't worry about passing down event handlers and trying to matchup selectors, but I believe this is the correct overall approach, and if we can make it work in the base case (where event handlers only select elements within their component) then we'll have the foundation for more complex--but rarer--use cases. It also follows the same pattern as helpers, i.e. where we assign them to the component. 

Another thing to note is that this will be very parallel to how both @mitar's and my component packages work. Not that this has anything to do with the 2.0 version--i think this is the way we should go even for the 1.0 version. It's the same logical considerations that drove me and @mitar to zap all the life cycle handlers, helpers and events on to a class. That same logic applies here where React Component's are classes. So it's a logical mapping that we should make early on in the code execution process. Then we can refer to all these stuff as methods from the component being rendered. This basically was already done in terms of `onCreated` becoming `componentWillMount`. So we need to continue that thought process. There's a reason why @TwinTailsX worked on things that seemed just like React Component/Class features. **_If we can make it seem like u can work on this by just adjusting methods on one main class OR regexes in a compiler we win!!!**_ That's our goal here. The plugin is "core" and we lock that down to allow for this pattern, then we move all remaining code into one regex file and one React Component class file.

Lastly, the React Component should be a mixin, rather than the methods defined on every component. We should have one file that looks like a plain old React Component. Then in all the generated components, u just specify it as a mixin OR inherit from it. My system eventually will allow u to inherit from `Meteor.Component` and u wont specify a mixin. It will just be automatic. Not even the meteor mixin that gives you `getMeteorData`. There's just no point in Meteor if ur always doing that. In this case, it might matter less since after all the code is generated. ..I do think for now for Blaze React, the best approach is not to do what I do and to mirror Meteor's current implementation, and just generate code that has 2:

```
 mixins: [ReactMeteorData, BlazeReact]
```

Hell Yea! ...the idea is eventually we wanna provide an optional feature to un the transpiler as a code generator, and have it export your code so u can use it yourself and u dont have to transpile anymore. So the code transpiled should look like a standard React Component rather than this:

```

Template = class {
  static _init(type, name) {
    Template[name] = class extends React.Component {
```

In Conclusion, more important than the tests, diagraming, is this refactoring. Only you can continue to work on it as is, but for any new developer it's 3x more complex than it needs to be understand. So I'd forget analyzing those compiler tests for now. Give your best shot at the ideas you like here and any of your own, and then we take it from there. At the very least that will up my participation.
